clustering_code/edchan-kmeans.qmd

```diff
@@ -0,0 +1,946 @@
+    blue_das = []
+    nir_das = []
+    swir1_das = []
+    swir2_das = []
+    merged_fmask = []
+    merged_red = []
+    merged_green = []
+    merged_blue = []
+    merged_nir = []
+    merged_swir1 = []
+    merged_swir2 = []
+    for band, gdf_by_band in gdf_by_date.groupby('band'):
+        # print(f'Processing band: {band}')
+        for index, row in gdf_by_band.iterrows():
+            filename = row.filename
+            folder = row.granule_id
+            filepath = os.path.join(cache_dir, folder, filename)
+            # print(filepath, row.band)
+            # Set masked=True to convert -9999 to NaN
+            if row.band == 'B02':
+                blue_da = rxr.open_rasterio(filepath, masked=True)
+                blue_da.name = 'B02_Blue'
+                blue_das.append(blue_da)
+            elif row.band == 'B03':
+                green_da = rxr.open_rasterio(filepath, masked=True)
+                green_da.name = 'B03_Green'
+                green_das.append(green_da)
+            elif row.band == 'B04':
+                red_da = rxr.open_rasterio(filepath, masked=True)
+                red_da.name = 'B04_Red'
+                red_das.append(red_da)
+            elif row.band == 'B05':
+                nir_da = rxr.open_rasterio(filepath, masked=True)
+                nir_da.name = 'B05_NIR'
+                nir_das.append(nir_da)
+            elif row.band == 'B06':
+                swir1_da = rxr.open_rasterio(filepath, masked=True)
+                swir1_da.name = 'B06_SWIR1'
+                swir1_das.append(swir1_da)
+            elif row.band == 'B07':
+                swir2_da = rxr.open_rasterio(filepath, masked=True)
+                swir2_da.name = 'B07_SWIR2'
+                swir2_das.append(swir2_da)
+            elif row.band == 'Fmask':
+                fmask_da = rxr.open_rasterio(filepath, masked=True)
+                fmask_das.append(fmask_da)
+    merged_blue = rxrmerge.merge_arrays(blue_das)
+    merged_green = rxrmerge.merge_arrays(green_das)
+    merged_red = rxrmerge.merge_arrays(red_das)
+    merged_nir = rxrmerge.merge_arrays(nir_das)
+    merged_swir1 = rxrmerge.merge_arrays(swir1_das)
+    merged_swir2 = rxrmerge.merge_arrays(swir2_das)
+    merged_fmask = rxrmerge.merge_arrays(fmask_das)
+
+    aoi = study_site.to_crs(merged_red.rio.crs)
+
+    clipped_fmask = merged_fmask.rio.clip_box(*aoi.total_bounds)
+    clipped_red = merged_red.rio.clip_box(*aoi.total_bounds)
+    clipped_green = merged_green.rio.clip_box(*aoi.total_bounds)
+    clipped_blue = merged_blue.rio.clip_box(*aoi.total_bounds)
+    clipped_nir = merged_nir.rio.clip_box(*aoi.total_bounds)
+    clipped_swir1 = merged_swir1.rio.clip_box(*aoi.total_bounds)
+    clipped_swir2 = merged_swir2.rio.clip_box(*aoi.total_bounds)
+
+    bits = np.unpackbits(clipped_fmask.values.astype(np.uint8)).reshape(clipped_fmask.shape + (-1,))
+    cloud_mask = np.prod(bits[..., mask_bits] == 0, axis=-1)
+
+    masked_red = clipped_red.where(cloud_mask)
```

Comment:
This section is kind of long and repetitive. Perhaps an alternative approach would be to apply the cloud mask in the ```for band```... loop, before the arrays are merged. Also, instead of creating a separate list for each band, you could create one list for each band in the ```for band``` loop, e.g.:

(pseudocode)

for date
&nbsp; for band ...
&nbsp;&nbsp; granules_das = [] 
&nbsp;&nbsp; for index... (I assume this is for each image/granule)
&nbsp;&nbsp;&nbsp; Apply cloud mask
&nbsp;&nbsp;&nbsp; Append the granules to the list
&nbsp; Merge arrays for the band
&nbsp; Clip the merged image for the band
...